# Battle of Neighbourhoods: A Case study of two Major Financial hubs in their Countries

## 1. Introduction

### 1.1 Background

The City of Johannesburg in South Africa is a major hub for economic activity in the country and is the financial centre of Africa’s most industrialised economy. Similar to Toronto in Canada, they are the financial capital centres as well as largest GDP contributors in their countries respectively. Toronto contributes about 20% to the GDP of Canada while Johannesburg contributes about 14%. Though the two countries are major contributors to the economies the size of the contribution varies vastly with GDP per capita of Toronto being 36000 US dollars and the GDP per capita for Johannesburg being 3000 US dollars GDP per capita. In a world where economies are fluid and international work opportunities are available for the skilled and mobile working professionals; as well as to investors looking to diversify their investments, understanding neighbourhoods in the two cities can be used as a marketing tool to help attract skilled contributors to the countries economy in the form of visitors, investors or as a way to boost local economy by the locals. 


### 1.2 Problem

The Financial Districts of Johannesburg and Toronto are by nature and design present a flurry of financial activities and attractions, stemming from the fact that the financial industry has been and continues to engine of growth and wealth in economies across the World. This study explores how an individual living or a  visitor to the two cities is likely to experience the neighbourhoods surrounding the two financial districts. The two financial districts will be indexed by the Johannesburg Stock Exchange and the Toronto Stock Exchange for Johannesburg and Toronot respectively. 


### 1.3 Interest

The study of the two financial districts would be of particular interest to, a working professional, looking to enrich their local experience (South African/Canadian), a visitor looking to enrich their international experience of South Africa or Toronto, or an investor conducting research to assess viability for a financial investment project. 

## 2 Data Acquistion and Cleaning

### 2.1 Data Acquisition

For the Initial starting points of the two financial districts the Google search engine was used to determine the address of the stock exchnage buildings. For the Toronto Stock Exchange, the Google search engine was further used to obtain the GPS coordinates of the building. For all data pertaining the project reliance was placed on the Foursquare API for input into the exploration and analysis work.

### 2.2 Feature Selection 

After extensive research and exploration of Foursquare API documentation, I settled on 3 API's to use for the study in this project. These are Building, Entertainment and Restaurant API’s. Selection of these API's in my view will be able to provide one with high level information and also directionally will be able to make educated decisions regarding neighbourhoods of the two financial districts. 

## 3. Data Exploration

As the study is an exploration study with reliance on on built in Python function. Neccesary libraries needed to perform analysis needed to be imported. The next section imports the relevant Python Libraries

In [79]:
import requests # library to handle requests
import pandas as pd # library for data analsysis
import numpy as np # library to handle data in a vectorized manner
import random # library for random number generation


!pip install geopy
from geopy.geocoders import Nominatim # module to convert an address into latitude and longitude values

# libraries for displaying images
from IPython.display import Image 
from IPython.core.display import HTML 
    
# tranforming json file into a pandas dataframe library
from pandas.io.json import json_normalize


! pip install folium==0.5.0
import folium # plotting library

%matplotlib inline 

import matplotlib as mpl
import matplotlib.pyplot as plt


print('Folium installed')
print('Libraries imported.')

  Using cached geopy-2.1.0-py3-none-any.whl (112 kB)
  Using cached geographiclib-1.52-py3-none-any.whl (38 kB)
Processing ./.cache/pip/wheels/b2/2f/2c/109e446b990d663ea5ce9b078b5e7c1a9c45cca91f377080f8/folium-0.5.0-py3-none-any.whl
  Using cached branca-0.4.2-py3-none-any.whl (24 kB)
  Using cached requests-2.25.1-py2.py3-none-any.whl (61 kB)
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
  Using cached Jinja2-3.0.1-py3-none-any.whl (133 kB)
  Using cached idna-2.10-py2.py3-none-any.whl (58 kB)
  Using cached certifi-2021.5.30-py2.py3-none-any.whl (145 kB)
  Using cached chardet-4.0.0-py2.py3-none-any.whl (178 kB)
  Using cached urllib3-1.26.6-py2.py3-none-any.whl (138 kB)
  Using cached MarkupSafe-2.0.1-cp37-cp37m-manylinux2010_x86_64.whl (31 kB)
ERROR: conda 4.8.2 requires ruamel_yaml>=0.11.14, which is not installed.
ERROR: tensorflow 2.1.0 has requirement scipy==1.4.1; python_version >= "3", but you'll have scipy 1.5.0 which is incompatible.
ERROR: botocore 1.16.11 has req

In [80]:
print ('Matplotlib version: ', mpl.__version__) # >= 2.0.0

Matplotlib version:  3.2.2


#### Adding Personal Credentials to use and Foursquare

In [2]:
CLIENT_ID = 'XP5V0UWY5UNHLA1PLLNKBCGTZCVKXNQIMFLB5BCTUMN3TXMC' # your Foursquare ID
CLIENT_SECRET = 'C0FGSOMMZL20PS3GZJ24N54TVVQAKMG1HKA1GL1YMPLMQPUU' # your Foursquare Secret
ACCESS_TOKEN = '1UUHMKB3ZBYQ3GVUNHWDRV3VFWZPBOT4YEWSTCO0EMXFYW2N' # your FourSquare Access Token

VERSION = '20180604'
LIMIT = 100
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: XP5V0UWY5UNHLA1PLLNKBCGTZCVKXNQIMFLB5BCTUMN3TXMC
CLIENT_SECRET:C0FGSOMMZL20PS3GZJ24N54TVVQAKMG1HKA1GL1YMPLMQPUU


## 3.1 Exploring Buildings around the Johannesburg Stock Exchange

Using the address obtained from the Google Search engine as input, the below program seeks to find the GPS coordinates of the Johannesburg stock exchnage building in South Africa. 

In [5]:
address = '2 Gwen Ln, Sandown, Sandton, 2196, South Africa'

geolocator = Nominatim(user_agent="foursquare_agent")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print(latitude, longitude)

-26.1021936 28.0565306


### 3.1.1 Determining Exploration Radius from the Johannesburg Stock Exchange

In [6]:
query = 'Building'
radius = 1000
print(query + ' .... OK!')

Building .... OK!


In [7]:
url_JSE = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude,ACCESS_TOKEN, VERSION, query, radius, LIMIT)


In [8]:
import requests

In [9]:
results_JSE = requests.get(url_JSE).json()


In [10]:
'There are {} Buildings around the Johannesburg Stock Exchange.'.format(len(results_JSE['response']['groups'][0]['items']))

'There are 6 Buildings around the Johannesburg Stock Exchange.'

### 3.1.2. Formating Results

In [11]:
items_JSE = results_JSE['response']['groups'][0]['items']


In [12]:
dataframe_JSE = pd.json_normalize(items_JSE) # flatten JSON

### 3.1.3. Displaying Summary Results 

In [13]:
dataframe_JSE.head()

referralId  reasons.count  \
0  e-0-4b05870df964a520547d22e3-0              0   
1  e-0-4bdf22f17ea362b52f5943c4-1              0   
2  e-0-4b06d37ef964a520f9f022e3-2              0   
3  e-0-4e0c8ee6d1646f1517874c2e-3              0   
4  e-0-4ce43f0e42f0b1f7b2448cce-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id                                venue.name  \
0  4b05870df964a520547d22e3                    Protea Hotel Balalaika   
1  4bdf22f17ea362b52f5943c4  Hydro Park on Grayston Apartment Sandton   
2  4b06d37ef964a520f9f022e3                              Sandton City   
3  4e0c8ee6d1646f1517874c2e                       Michelangelo Towers   
4  4ce43f0e42f0b1f7b2448cce                    Wall Street Restaurant   

  venue.location.address  venue.location.lat  venue.location.lng  \
0  Maude Street, Sandown          -26.103978           28.058131   
1      86 Grayston Drive          -26.099973           28.053730   
2             Rivonia Rd          -26.108566           28.052848   
3         Sandown Valley          -26.106438           28.054180   
4                    NaN          -26.102887           28.060970   

                       venue.location.labeledLatLngs  venue.location.distance  \
0  [{'label': 'display', 'lat': -26.1039778, 'lng...                      255   
1  [{'label': 'display', 'lat': -26.0999725344602...                      373   
2                                                NaN                      799   
3  [{'label': 'display', 'lat': -26.1064382621836...                      527   
4  [{'label': 'display', 'lat': -26.1028865600020...                      450   

   ... venue.location.cc venue.location.neighborhood venue.location.city  \
0  ...                ZA                     Sandown             Sandton   
1  ...                ZA                         NaN             Sandton   
2  ...                ZA                         NaN             Sandton   
3  ...                ZA                         NaN             Sandton   
4  ...                ZA                         NaN                 NaN   

  venue.location.state venue.location.country  \
0                ZA-GT      iNingizimu Afrika   
1             IGauteng      iNingizimu Afrika   
2             IGauteng      iNingizimu Afrika   
3             IGauteng      iNingizimu Afrika   
4                  NaN      iNingizimu Afrika   

                venue.location.formattedAddress  \
0        [Maude Street, Sandown, Sandton, 2126]   
1                  [86 Grayston Drive, Sandton]   
2  [Rivonia Rd (cnr Sandton Dr), Sandton, 2196]   
3      [Sandown Valley (Maude street), Sandton]   
4                                           NaN   

                                    venue.categories venue.photos.count  \
0  [{'id': '4bf58dd8d48988d1fa931735', 'name': 'H...                  0   
1  [{'id': '4bf58dd8d48988d1fa931735', 'name': 'H...                  0   
2  [{'id': '4bf58dd8d48988d1fd941735', 'name': 'S...                  0   
3  [{'id': '4bf58dd8d48988d1fa931735', 'name': 'H...                  0   
4  [{'id': '4bf58dd8d48988d1c8941735', 'name': 'A...                  0   

   venue.photos.groups venue.location.crossStreet  
0                   []                        NaN  
1                   []                        NaN  
2                   []             cnr Sandton Dr  
3                   []               Maude street  
4                   []                        NaN  

[5 rows x 21 columns]

In [14]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe_JSE.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered_JSE = dataframe_JSE.loc[:, filtered_columns]


# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered_JSE['venue.categories'] = dataframe_filtered_JSE.apply(get_category_type, axis=1)


# clean columns
dataframe_filtered_JSE.columns = [col.split('.')[-1] for col in dataframe_filtered_JSE.columns]

dataframe_filtered_JSE.head(10)

name          categories  \
0                    Protea Hotel Balalaika               Hotel   
1  Hydro Park on Grayston Apartment Sandton               Hotel   
2                              Sandton City       Shopping Mall   
3                       Michelangelo Towers               Hotel   
4                    Wall Street Restaurant  African Restaurant   
5                          InterContinental               Hotel   

                        address        lat        lng  \
0         Maude Street, Sandown -26.103978  28.058131   
1             86 Grayston Drive -26.099973  28.053730   
2                    Rivonia Rd -26.108566  28.052848   
3                Sandown Valley -26.106438  28.054180   
4                           NaN -26.102887  28.060970   
5  Corner of Maude & 5th Street -26.106805  28.052569   

                                      labeledLatLngs  distance postalCode  cc  \
0  [{'label': 'display', 'lat': -26.1039778, 'lng...       255       2126  ZA   
1  [{'label': 'display', 'lat': -26.0999725344602...       373        NaN  ZA   
2                                                NaN       799       2196  ZA   
3  [{'label': 'display', 'lat': -26.1064382621836...       527        NaN  ZA   
4  [{'label': 'display', 'lat': -26.1028865600020...       450        NaN  ZA   
5  [{'label': 'display', 'lat': -26.1068045365651...       648       2146  ZA   

  neighborhood     city     state            country  \
0      Sandown  Sandton     ZA-GT  iNingizimu Afrika   
1          NaN  Sandton  IGauteng  iNingizimu Afrika   
2          NaN  Sandton  IGauteng  iNingizimu Afrika   
3          NaN  Sandton  IGauteng  iNingizimu Afrika   
4          NaN      NaN       NaN  iNingizimu Afrika   
5          NaN  Sandton  IGauteng  iNingizimu Afrika   

                                formattedAddress     crossStreet  \
0         [Maude Street, Sandown, Sandton, 2126]             NaN   
1                   [86 Grayston Drive, Sandton]             NaN   
2   [Rivonia Rd (cnr Sandton Dr), Sandton, 2196]  cnr Sandton Dr   
3       [Sandown Valley (Maude street), Sandton]    Maude street   
4                                            NaN             NaN   
5  [Corner of Maude & 5th Street, Sandton, 2146]             NaN   

                         id  
0  4b05870df964a520547d22e3  
1  4bdf22f17ea362b52f5943c4  
2  4b06d37ef964a520f9f022e3  
3  4e0c8ee6d1646f1517874c2e  
4  4ce43f0e42f0b1f7b2448cce  
5  4b05870df964a5202b7d22e3

In [15]:
print(dataframe_filtered_JSE.columns.tolist())

['name', 'categories', 'address', 'lat', 'lng', 'labeledLatLngs', 'distance', 'postalCode', 'cc', 'neighborhood', 'city', 'state', 'country', 'formattedAddress', 'crossStreet', 'id']


### 3.1.4 Displaying a Formatted Data points of interests from the Results

In [16]:
dataframe_2 = dataframe_filtered_JSE.iloc [:, 0:5]

In [17]:
dataframe_2

name          categories  \
0                    Protea Hotel Balalaika               Hotel   
1  Hydro Park on Grayston Apartment Sandton               Hotel   
2                              Sandton City       Shopping Mall   
3                       Michelangelo Towers               Hotel   
4                    Wall Street Restaurant  African Restaurant   
5                          InterContinental               Hotel   

                        address        lat        lng  
0         Maude Street, Sandown -26.103978  28.058131  
1             86 Grayston Drive -26.099973  28.053730  
2                    Rivonia Rd -26.108566  28.052848  
3                Sandown Valley -26.106438  28.054180  
4                           NaN -26.102887  28.060970  
5  Corner of Maude & 5th Street -26.106805  28.052569

## 3.2 Exploring Buildings around Toronto Stock Exchange

### 3.2.1 Adding Toronto Stock Exchange Address

In [18]:
address = '130 King St W, Toronto, ON M5X 2A2, Canada'

geolocator = Nominatim(user_agent="foursquare_agent")
location =  geolocator.geocode(address)
latitude_2 = 43.648550 #location.latitude
longitude_2 = -79.383304 #location.longitude
print(latitude_2, longitude_2)

43.64855 -79.383304


In [19]:
query = 'Building'
radius_1 = 1000
print(query + ' .... OK!')

Building .... OK!


In [20]:
url_TSX = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude_2, longitude_2,ACCESS_TOKEN, VERSION, query, radius_1, LIMIT)


### 3.2.2. Formating Results

In [21]:
import requests

In [22]:
results_TSX = requests.get(url_TSX).json()


In [23]:
'There are {} buildings around the Toronto Stock Exchange.'.format(len(results_TSX['response']['groups'][0]['items']))

'There are 65 buildings around the Toronto Stock Exchange.'

In [24]:
items_TSX = results_TSX['response']['groups'][0]['items']

In [25]:
dataframe_TSX = pd.json_normalize(items_TSX) # flatten JSON


### 3.2.3. Displaying Results 

In [26]:
dataframe_TSX.head()

referralId  reasons.count  \
0  e-0-4ad4c05ef964a5208ef620e3-0              0   
1  e-0-4adb896df964a520bf2821e3-1              0   
2  e-0-4ad4c05ef964a520daf620e3-2              0   
3  e-0-4bcf714ab221c9b67f0ad2d0-3              0   
4  e-0-4ad4c05ef964a52096f620e3-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id              venue.name venue.location.address  \
0  4ad4c05ef964a5208ef620e3           Old City Hall   60 Queen Street West   
1  4adb896df964a520bf2821e3            401 Richmond      401 Richmond St W   
2  4ad4c05ef964a520daf620e3  Art Gallery of Ontario        317 Dundas St W   
3  4bcf714ab221c9b67f0ad2d0      TIFF Bell Lightbox         350 King St. W   
4  4ad4c05ef964a52096f620e3                CN Tower         301 Front St W   

   venue.location.lat  venue.location.lng  \
0           43.652009          -79.381744   
1           43.647906          -79.395219   
2           43.654003          -79.392922   
3           43.646342          -79.390063   
4           43.642277          -79.387055   

                       venue.location.labeledLatLngs  venue.location.distance  \
0  [{'label': 'display', 'lat': 43.65200880087612...                      405   
1  [{'label': 'display', 'lat': 43.64790630817995...                      962   
2  [{'label': 'display', 'lat': 43.65400286033738...                      984   
3  [{'label': 'display', 'lat': 43.64634159071232...                      597   
4                                                NaN                      760   

   ... venue.location.city venue.location.state venue.location.country  \
0  ...             Toronto                   ON                 Canada   
1  ...             Toronto                   ON                 Canada   
2  ...             Toronto                   ON                 Canada   
3  ...             Toronto                   ON                 Canada   
4  ...             Toronto                   ON                 Canada   

                     venue.location.formattedAddress  \
0  [60 Queen Street West, Toronto ON M5H 1A1, Can...   
1  [401 Richmond St W (at Spadina Ave), Toronto O...   
2  [317 Dundas St W (at Beverley St), Toronto ON ...   
3  [350 King St. W (at John St.), Toronto ON M5V ...   
4  [301 Front St W (at John St), Toronto ON M5V 2...   

                                    venue.categories venue.photos.count  \
0  [{'id': '4bf58dd8d48988d12d941735', 'name': 'M...                  0   
1  [{'id': '4bf58dd8d48988d1e2931735', 'name': 'A...                  0   
2  [{'id': '4bf58dd8d48988d1e2931735', 'name': 'A...                  0   
3  [{'id': '4bf58dd8d48988d17f941735', 'name': 'M...                  0   
4  [{'id': '4bf58dd8d48988d12d941735', 'name': 'M...                  0   

  venue.photos.groups  venue.location.crossStreet venue.venuePage.id  \
0                  []                         NaN                NaN   
1                  []              at Spadina Ave           33863999   
2                  []              at Beverley St           33853777   
3                  []                 at John St.           34450648   
4                  []                  at John St           33864000   

  venue.location.neighborhood  
0                         NaN  
1                         NaN  
2                         NaN  
3      Entertainment District  
4      Entertainment District  

[5 rows x 22 columns]

In [27]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe_TSX.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered_TSX = dataframe_TSX.loc[:, filtered_columns]


# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered_TSX['venue.categories'] = dataframe_filtered_TSX.apply(get_category_type, axis=1)


# clean columns
dataframe_filtered_TSX.columns = [col.split('.')[-1] for col in dataframe_filtered_TSX.columns]

dataframe_filtered_TSX.head(10)

name           categories  \
0                                Old City Hall  Monument / Landmark   
1                                 401 Richmond          Art Gallery   
2                       Art Gallery of Ontario          Art Gallery   
3                           TIFF Bell Lightbox        Movie Theater   
4                                     CN Tower  Monument / Landmark   
5                       Nathan Phillips Square                Plaza   
6  Four Seasons Centre for the Performing Arts         Concert Hall   
7                             Roy Thomson Hall         Concert Hall   
8                              Design Exchange          Art Gallery   
9                         First Canadian Place             Building   

                address        lat        lng  \
0  60 Queen Street West  43.652009 -79.381744   
1     401 Richmond St W  43.647906 -79.395219   
2       317 Dundas St W  43.654003 -79.392922   
3        350 King St. W  43.646342 -79.390063   
4        301 Front St W  43.642277 -79.387055   
5        100 Queen St W  43.652270 -79.383516   
6       145 Queen St. W  43.650592 -79.385806   
7          60 Simcoe St  43.646589 -79.385979   
8           234 Bay St.  43.647972 -79.380104   
9         100 King St W  43.648482 -79.382443   

                                      labeledLatLngs  distance postalCode  cc  \
0  [{'label': 'display', 'lat': 43.65200880087612...       405    M5H 1A1  CA   
1  [{'label': 'display', 'lat': 43.64790630817995...       962    M5V 3A8  CA   
2  [{'label': 'display', 'lat': 43.65400286033738...       984    M5T 1G4  CA   
3  [{'label': 'display', 'lat': 43.64634159071232...       597    M5V 3X5  CA   
4                                                NaN       760    M5V 2T6  CA   
5  [{'label': 'display', 'lat': 43.65227047322295...       414    M5H 2N1  CA   
6  [{'label': 'display', 'lat': 43.650592, 'lng':...       303    M5H 4G1  CA   
7  [{'label': 'display', 'lat': 43.6465886970634,...       306    M5J 2H5  CA   
8  [{'label': 'display', 'lat': 43.64797219581623...       265    M5K 1B2  CA   
9  [{'label': 'display', 'lat': 43.64848212292913...        69    M5J 1W9  CA   

      city state country                                   formattedAddress  \
0  Toronto    ON  Canada  [60 Queen Street West, Toronto ON M5H 1A1, Can...   
1  Toronto    ON  Canada  [401 Richmond St W (at Spadina Ave), Toronto O...   
2  Toronto    ON  Canada  [317 Dundas St W (at Beverley St), Toronto ON ...   
3  Toronto    ON  Canada  [350 King St. W (at John St.), Toronto ON M5V ...   
4  Toronto    ON  Canada  [301 Front St W (at John St), Toronto ON M5V 2...   
5  Toronto    ON  Canada  [100 Queen St W (at Bay St), Toronto ON M5H 2N...   
6  Toronto    ON  Canada  [145 Queen St. W (at University Ave.), Toronto...   
7  Toronto    ON  Canada  [60 Simcoe St (at King St W), Toronto ON M5J 2...   
8  Toronto    ON  Canada  [234 Bay St. (btwn King & Wellington), Toronto...   
9  Toronto    ON  Canada  [100 King St W (btwn Bay & University), Toront...   

              crossStreet            neighborhood                        id  
0                     NaN                     NaN  4ad4c05ef964a5208ef620e3  
1          at Spadina Ave                     NaN  4adb896df964a520bf2821e3  
2          at Beverley St                     NaN  4ad4c05ef964a520daf620e3  
3             at John St.  Entertainment District  4bcf714ab221c9b67f0ad2d0  
4              at John St  Entertainment District  4ad4c05ef964a52096f620e3  
5               at Bay St                     NaN  4ad4c05ef964a520a6f620e3  
6      at University Ave.                     NaN  4ad4c062f964a520e5f720e3  
7            at King St W                     NaN  4ad4c061f964a520b0f720e3  
8  btwn King & Wellington                     NaN  4ad4c05ff964a520eff620e3  
9   btwn Bay & University                     NaN  4ad4c063f964a52020f820e3

### 3.2.4 Displaying a Formatted Data points of interests from the Results

In [28]:
dataframe_filtered_TSX_1 = dataframe_filtered_TSX.iloc [:,0:5]

In [29]:
dataframe_filtered_TSX_1

name                categories  \
0                 Old City Hall       Monument / Landmark   
1                  401 Richmond               Art Gallery   
2        Art Gallery of Ontario               Art Gallery   
3            TIFF Bell Lightbox             Movie Theater   
4                      CN Tower       Monument / Landmark   
..                          ...                       ...   
60              Storys Building                Restaurant   
61         Union Station Market                    Market   
62  Loft404 Inspired Gatherings               Event Space   
63    Union Station Platform 27                  Platform   
64                Feta & Olives  Mediterranean Restaurant   

                  address        lat        lng  
0    60 Queen Street West  43.652009 -79.381744  
1       401 Richmond St W  43.647906 -79.395219  
2         317 Dundas St W  43.654003 -79.392922  
3          350 King St. W  43.646342 -79.390063  
4          301 Front St W  43.642277 -79.387055  
..                    ...        ...        ...  
60          11 Duncan St.  43.647552 -79.388318  
61                    NaN  43.645630 -79.381130  
62  263 Adelaide St. West  43.647851 -79.389490  
63         65 Front St. W  43.643952 -79.380496  
64        Pinnacle Centre  43.643451 -79.375566  

[65 rows x 5 columns]

In [30]:
Building_categories = dataframe_filtered_TSX_1['categories'].unique()

### 3.2.5 Creating a Frequency Table for the Data points of interests from the Results

In [31]:
TSX_Building_categories =  dataframe_filtered_TSX_1.groupby('categories')['categories'].size().reset_index(name='counts').sort_values(['counts','categories'],ascending = [False,True])

In [32]:
TSX_Building_categories

categories  counts
20                                     Hotel       8
11                               Coffee Shop       4
0                                Art Gallery       3
12                              Concert Hall       3
14                          Department Store       2
15                               Event Space       2
16                            Farmers Market       2
28                       Monument / Landmark       2
29                             Movie Theater       2
31                                    Office       2
38                                Steakhouse       2
40                                   Theater       2
1                                       Bank       1
2                           Baseball Stadium       1
3                         Basketball Stadium       1
4                         Belgian Restaurant       1
5                                    Brewery       1
6                                   Building       1
7                                Bus Station       1
8                                       Café       1
9                                     Church       1
10                            Clothing Store       1
13                             Deli / Bodega       1
17                                Food Court       1
18                                 Gastropub       1
19                            General Travel       1
21                               IT Services       1
22                        Italian Restaurant       1
23                              Liquor Store       1
24                                    Lounge       1
25                                    Market       1
26                  Mediterranean Restaurant       1
27                             Metro Station       1
30                                    Museum       1
32                                  Platform       1
33                                     Plaza       1
34  Residential Building (Apartment / Condo)       1
35                                Restaurant       1
36                             Shopping Mall       1
37                                Sports Bar       1
39                          Sushi Restaurant       1
41                             Train Station       1
42                                University       1

## 3.3 Exploring Entertainment spots around the Financial Districts

### 3.3.1 Exploring Entertainment spots around the Financial District in Johannesburg (Johannesburg Stock Exchange)

In [33]:
query = 'Entertainment'
radius_1 = 1000
print(query + ' .... OK!')

Entertainment .... OK!


In [34]:
url_JSE_fun = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude,ACCESS_TOKEN, VERSION, query, radius, LIMIT)


### 3.3.2. Formatting Results for Entertainment spots around the Johannesburg Stock Exchange 

In [35]:
results_JSE_fun = requests.get(url_JSE_fun).json()


In [36]:
'There are {} areas of enterainment around the Johannesburg Stock Exchange.'.format(len(results_JSE_fun['response']['groups'][0]['items']))

'There are 4 areas of enterainment around the Johannesburg Stock Exchange.'

In [37]:
items_JSE_fun = results_JSE_fun['response']['groups'][0]['items']


In [38]:
dataframe_JSE_fun = pd.json_normalize(items_JSE_fun) # flatten JSON

### 3.3.3. Displaying Results for Entertainment spots around the Johannesburg Stock Exchange  

In [39]:
dataframe_JSE_fun.head()

referralId  reasons.count  \
0  e-0-4bc9b94a937ca593c82ca692-0              0   
1  e-0-4bc10821f8219c74e1bab210-1              0   
2  e-0-50aa5d8ae4b0a68d85ecd45b-2              0   
3  e-0-4ee32b25be7b4845a30985a6-3              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id                   venue.name  \
0  4bc9b94a937ca593c82ca692      Harley Davidson Sandton   
1  4bc10821f8219c74e1bab210                        Taboo   
2  50aa5d8ae4b0a68d85ecd45b                    The Venue   
3  4ee32b25be7b4845a30985a6  Sandton City Fountain Court   

  venue.location.address venue.location.crossStreet  venue.location.lat  \
0       167 Rivonia road                   Grayston          -26.095784   
1             24 Central   Cnr Fredman Dr & Gwen Ln          -26.101860   
2             Green Park                 Lower Road          -26.096407   
3                    NaN                        NaN          -26.107452   

   venue.location.lng                      venue.location.labeledLatLngs  \
0           28.056061  [{'label': 'display', 'lat': -26.0957839571620...   
1           28.057470  [{'label': 'display', 'lat': -26.1018604279674...   
2           28.053734  [{'label': 'display', 'lat': -26.0964066688268...   
3           28.054710  [{'label': 'display', 'lat': -26.1074522861225...   

   venue.location.distance venue.location.cc venue.location.city  \
0                      715                ZA             Sandton   
1                      100                ZA             Sandton   
2                      702                ZA             Sandton   
3                      613                ZA                 NaN   

  venue.location.state venue.location.country  \
0             IGauteng      iNingizimu Afrika   
1             IGauteng      iNingizimu Afrika   
2             IGauteng      iNingizimu Afrika   
3                  NaN      iNingizimu Afrika   

                     venue.location.formattedAddress  \
0             [167 Rivonia road (Grayston), Sandton]   
1  [24 Central (Cnr Fredman Dr & Gwen Ln), Sandto...   
2                 [Green Park (Lower Road), Sandton]   
3                                                NaN   

                                    venue.categories  venue.photos.count  \
0  [{'id': '4bf58dd8d48988d1f1931735', 'name': 'G...                   0   
1  [{'id': '4bf58dd8d48988d11f941735', 'name': 'N...                   0   
2  [{'id': '4bf58dd8d48988d1f1931735', 'name': 'G...                   0   
3  [{'id': '4bf58dd8d48988d1f1931735', 'name': 'G...                   0   

  venue.photos.groups venue.location.postalCode  
0                  []                       NaN  
1                  []                      2196  
2                  []                       NaN  
3                  []                       NaN

In [40]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe_JSE_fun.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered_JSE_fun = dataframe_JSE_fun.loc[:, filtered_columns]


# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered_JSE_fun['venue.categories'] = dataframe_filtered_JSE_fun.apply(get_category_type, axis=1)


# clean columns
dataframe_filtered_JSE_fun.columns = [col.split('.')[-1] for col in dataframe_filtered_JSE_fun.columns]

dataframe_filtered_JSE_fun.head(10)

name             categories           address  \
0      Harley Davidson Sandton  General Entertainment  167 Rivonia road   
1                        Taboo              Nightclub        24 Central   
2                    The Venue  General Entertainment        Green Park   
3  Sandton City Fountain Court  General Entertainment               NaN   

                crossStreet        lat        lng  \
0                  Grayston -26.095784  28.056061   
1  Cnr Fredman Dr & Gwen Ln -26.101860  28.057470   
2                Lower Road -26.096407  28.053734   
3                       NaN -26.107452  28.054710   

                                      labeledLatLngs  distance  cc     city  \
0  [{'label': 'display', 'lat': -26.0957839571620...       715  ZA  Sandton   
1  [{'label': 'display', 'lat': -26.1018604279674...       100  ZA  Sandton   
2  [{'label': 'display', 'lat': -26.0964066688268...       702  ZA  Sandton   
3  [{'label': 'display', 'lat': -26.1074522861225...       613  ZA      NaN   

      state            country  \
0  IGauteng  iNingizimu Afrika   
1  IGauteng  iNingizimu Afrika   
2  IGauteng  iNingizimu Afrika   
3       NaN  iNingizimu Afrika   

                                    formattedAddress postalCode  \
0             [167 Rivonia road (Grayston), Sandton]        NaN   
1  [24 Central (Cnr Fredman Dr & Gwen Ln), Sandto...       2196   
2                 [Green Park (Lower Road), Sandton]        NaN   
3                                                NaN        NaN   

                         id  
0  4bc9b94a937ca593c82ca692  
1  4bc10821f8219c74e1bab210  
2  50aa5d8ae4b0a68d85ecd45b  
3  4ee32b25be7b4845a30985a6

### 3.3.4. Displaying Data points of the Interest from the results set for Entertainment spots around the Johannesburg Stock Exchange 

In [41]:
dataframe_JSE_entertain = dataframe_filtered_JSE_fun.iloc [:, 0:5]

In [42]:
dataframe_JSE_entertain

name             categories           address  \
0      Harley Davidson Sandton  General Entertainment  167 Rivonia road   
1                        Taboo              Nightclub        24 Central   
2                    The Venue  General Entertainment        Green Park   
3  Sandton City Fountain Court  General Entertainment               NaN   

                crossStreet        lat  
0                  Grayston -26.095784  
1  Cnr Fredman Dr & Gwen Ln -26.101860  
2                Lower Road -26.096407  
3                       NaN -26.107452

In [43]:
Entertainment_Categories_JHB = dataframe_JSE_entertain.groupby('categories')['categories'].size().reset_index(name='counts').sort_values(['counts','categories'],ascending = [False,True])
Entertainment_Categories_JHB

categories  counts
0  General Entertainment       3
1              Nightclub       1

## 3.4. Exploring Entertainment spots around the Financial District in Toronto (Toronto Stock Exchange)

### 3.4.1. Executing the API for Exploring Entertainment spots around the Toronto Stock Exchange 

In [44]:
url_TSX_fun = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude_2, longitude_2,ACCESS_TOKEN, VERSION, query, radius_1, LIMIT)


### 3.4.2. Formatting Results for Entertainment spots around the Toronto Stock Exchange 

In [45]:
results_TSX_fun = requests.get(url_TSX_fun).json()
results_TSX_fun

{'meta': {'code': 200, 'requestId': '60dd2ccdb5ab84411e701bfb'},
 'notifications': [{'type': 'notificationTray', 'item': {'unreadCount': 0}}],
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Financial District',
  'headerFullLocation': 'Financial District, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'query': 'entertainment',
  'totalResults': 42,
  'suggestedBounds': {'ne': {'lat': 43.65755000900001,
    'lng': -79.37088918065864},
   'sw': {'lat': 43.63954999099999, 'lng': -79.39571881934135}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5033a657e4b0c160f748c5ab',
       'name': 'TIFF Festival Box Office',
       'location': {'address': '225 King St. W',
       

In [46]:
'There are {} areas of enterainment around the Toronto Stock Exchange.'.format(len(results_TSX_fun['response']['groups'][0]['items']))

'There are 42 areas of enterainment around the Toronto Stock Exchange.'

In [47]:
items_TSX_fun = results_TSX_fun['response']['groups'][0]['items']


In [48]:
dataframe_TSX_fun = pd.json_normalize(items_TSX_fun) # flatten JSON

In [49]:
dataframe_TSX_fun.head()

referralId  reasons.count  \
0  e-0-5033a657e4b0c160f748c5ab-0              0   
1  e-0-4bd32bdea8b3a593de3e695f-1              0   
2  e-0-4ad4c062f964a520b5f720e3-2              0   
3  e-0-4ad4c061f964a520aef720e3-3              0   
4  e-0-4ad4c05ef964a52096f620e3-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id                        venue.name  \
0  5033a657e4b0c160f748c5ab          TIFF Festival Box Office   
1  4bd32bdea8b3a593de3e695f                      FITC Toronto   
2  4ad4c062f964a520b5f720e3  Elgin And Winter Garden Theatres   
3  4ad4c061f964a520aef720e3         Princess Of Wales Theatre   
4  4ad4c05ef964a52096f620e3                          CN Tower   

  venue.location.address   venue.location.crossStreet  venue.location.lat  \
0         225 King St. W              in Metro Centre           43.646887   
1     145 Richmond St. W           at University Ave.           43.649872   
2           189 Yonge St  btwn Queen St E & Shuter St           43.653394   
3         300 King St. W                  at John St.           43.646543   
4         301 Front St W                   at John St           43.642277   

   venue.location.lng                      venue.location.labeledLatLngs  ...  \
0          -79.387885  [{'label': 'display', 'lat': 43.64688731930244...  ...   
1          -79.385745  [{'label': 'display', 'lat': 43.64987248299678...  ...   
2          -79.378507  [{'label': 'display', 'lat': 43.65339379601958...  ...   
3          -79.389100  [{'label': 'display', 'lat': 43.64654315405365...  ...   
4          -79.387055                                                NaN  ...   

   venue.location.cc venue.location.city venue.location.state  \
0                 CA             Toronto                   ON   
1                 CA             Toronto                   ON   
2                 CA             Toronto                   ON   
3                 CA             Toronto                   ON   
4                 CA             Toronto                   ON   

  venue.location.country                    venue.location.formattedAddress  \
0                 Canada  [225 King St. W (in Metro Centre), Toronto ON ...   
1                 Canada  [145 Richmond St. W (at University Ave.), Toro...   
2                 Canada  [189 Yonge St (btwn Queen St E & Shuter St), T...   
3                 Canada  [300 King St. W (at John St.), Toronto ON M5V ...   
4                 Canada  [301 Front St W (at John St), Toronto ON M5V 2...   

                                    venue.categories venue.photos.count  \
0  [{'id': '4bf58dd8d48988d1f1931735', 'name': 'G...                  0   
1  [{'id': '4bf58dd8d48988d1f1931735', 'name': 'G...                  0   
2  [{'id': '4bf58dd8d48988d137941735', 'name': 'T...                  0   
3  [{'id': '4bf58dd8d48988d1f1931735', 'name': 'G...                  0   
4  [{'id': '4bf58dd8d48988d12d941735', 'name': 'M...                  0   

  venue.photos.groups  venue.location.neighborhood venue.venuePage.id  
0                  []                          NaN                NaN  
1                  []                          NaN                NaN  
2                  []                          NaN                NaN  
3                  []                          NaN                NaN  
4                  []       Entertainment District           33864000  

[5 rows x 22 columns]

### 3.4.3. Displaying Results for Entertainment spots around the Toronto Stock Exchange 

In [50]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe_TSX_fun.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered_TSX_fun = dataframe_TSX_fun.loc[:, filtered_columns]


# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered_TSX_fun['venue.categories'] = dataframe_filtered_TSX_fun.apply(get_category_type, axis=1)


# clean columns
dataframe_filtered_TSX_fun.columns = [col.split('.')[-1] for col in dataframe_filtered_TSX_fun.columns]

dataframe_filtered_TSX_fun.head(10)

name             categories  \
0                     TIFF Festival Box Office  General Entertainment   
1                                 FITC Toronto  General Entertainment   
2             Elgin And Winter Garden Theatres                Theater   
3                    Princess Of Wales Theatre  General Entertainment   
4                                     CN Tower    Monument / Landmark   
5                      Royal Alexandra Theatre  General Entertainment   
6                       Albany Club of Toronto  General Entertainment   
7  Four Seasons Centre for the Performing Arts           Concert Hall   
8                           TIFF Bell Lightbox          Movie Theater   
9                             Scotiabank Arena     Basketball Stadium   

              address                  crossStreet        lat        lng  \
0      225 King St. W              in Metro Centre  43.646887 -79.387885   
1  145 Richmond St. W           at University Ave.  43.649872 -79.385745   
2        189 Yonge St  btwn Queen St E & Shuter St  43.653394 -79.378507   
3      300 King St. W                  at John St.  43.646543 -79.389100   
4      301 Front St W                   at John St  43.642277 -79.387055   
5       260 King St W                 at Duncan St  43.647255 -79.387530   
6       91 King St. E                at Church St.  43.649564 -79.374923   
7     145 Queen St. W           at University Ave.  43.650592 -79.385806   
8      350 King St. W                  at John St.  43.646342 -79.390063   
9           40 Bay St                          NaN  43.643446 -79.379040   

                                      labeledLatLngs  distance postalCode  cc  \
0  [{'label': 'display', 'lat': 43.64688731930244...       412    M5V 3C7  CA   
1  [{'label': 'display', 'lat': 43.64987248299678...       245    M5H 2L2  CA   
2  [{'label': 'display', 'lat': 43.65339379601958...       663    M5B 2H1  CA   
3  [{'label': 'display', 'lat': 43.64654315405365...       517    M5V 1J2  CA   
4                                                NaN       760    M5V 2T6  CA   
5  [{'label': 'display', 'lat': 43.647255, 'lng':...       369    M5V 1H9  CA   
6  [{'label': 'display', 'lat': 43.64956374719146...       684    M5C 1G3  CA   
7  [{'label': 'display', 'lat': 43.650592, 'lng':...       303    M5H 4G1  CA   
8  [{'label': 'display', 'lat': 43.64634159071232...       597    M5V 3X5  CA   
9  [{'label': 'display', 'lat': 43.64344617535107...       663    M5J 2X2  CA   

      city state country                                   formattedAddress  \
0  Toronto    ON  Canada  [225 King St. W (in Metro Centre), Toronto ON ...   
1  Toronto    ON  Canada  [145 Richmond St. W (at University Ave.), Toro...   
2  Toronto    ON  Canada  [189 Yonge St (btwn Queen St E & Shuter St), T...   
3  Toronto    ON  Canada  [300 King St. W (at John St.), Toronto ON M5V ...   
4  Toronto    ON  Canada  [301 Front St W (at John St), Toronto ON M5V 2...   
5  Toronto    ON  Canada  [260 King St W (at Duncan St), Toronto ON M5V ...   
6  Toronto    ON  Canada  [91 King St. E (at Church St.), Toronto ON M5C...   
7  Toronto    ON  Canada  [145 Queen St. W (at University Ave.), Toronto...   
8  Toronto    ON  Canada  [350 King St. W (at John St.), Toronto ON M5V ...   
9  Toronto    ON  Canada            [40 Bay St, Toronto ON M5J 2X2, Canada]   

             neighborhood                        id  
0                     NaN  5033a657e4b0c160f748c5ab  
1                     NaN  4bd32bdea8b3a593de3e695f  
2                     NaN  4ad4c062f964a520b5f720e3  
3                     NaN  4ad4c061f964a520aef720e3  
4  Entertainment District  4ad4c05ef964a52096f620e3  
5  Entertainment District  4ad4c062f964a520bbf720e3  
6                     NaN  4b6a001df964a520dbc02be3  
7                     NaN  4ad4c062f964a520e5f720e3  
8  Entertainment District  4bcf714ab221c9b67f0ad2d0  
9                     NaN  4b155081f964a520b4b023e3

### 3.4.4. Displaying Data points of interest from the Results for Entertainment spots around the Toronto Stock Exchange 

In [51]:
dataframe_filtered_TSX_fun_1 = dataframe_filtered_TSX_fun.iloc [:,0:5]

In [52]:
dataframe_filtered_TSX_fun_1

name             categories  \
0                            TIFF Festival Box Office  General Entertainment   
1                                        FITC Toronto  General Entertainment   
2                    Elgin And Winter Garden Theatres                Theater   
3                           Princess Of Wales Theatre  General Entertainment   
4                                            CN Tower    Monument / Landmark   
5                             Royal Alexandra Theatre  General Entertainment   
6                              Albany Club of Toronto  General Entertainment   
7         Four Seasons Centre for the Performing Arts           Concert Hall   
8                                  TIFF Bell Lightbox          Movie Theater   
9                                    Scotiabank Arena     Basketball Stadium   
10              LG Fashion Week (David Pecaut Square)  General Entertainment   
11                    Toronto Maple Leafs Hockey Club  General Entertainment   
12                               JL Recording Studios  General Entertainment   
13                                Cavalcade of Lights  General Entertainment   
14                TwestivalTO @ The Fifth Social Club  General Entertainment   
15                                                CTV  General Entertainment   
16                                               CP24  General Entertainment   
17  BRIDGE BETWEEN NORTH & SOUTH BLDG - MTCC - Lev...  General Entertainment   
18                        Loft404 Inspired Gatherings            Event Space   
19                                         The Social  General Entertainment   
20                             The Marilyn Denis Show  General Entertainment   
21                                  MuchMusic Studios  General Entertainment   
22                                          6ix Still  General Entertainment   
23             Kinect for Xbox 360 Experience Toronto  General Entertainment   
24                                    Comicon Toronto  General Entertainment   
25                 EXHIBIT HALL G - MTCC - South Bldg  General Entertainment   
26                       The Score Television Network  General Entertainment   
27                          VISA Screening Room #TIFF          Movie Theater   
28                                          The Porch                    Bar   
29                                       The Rec Room                    Bar   
30                                       Juniper Park  General Entertainment   
31                                               NXNE  General Entertainment   
32                                David Pecaut Square                   Park   
33                                            Captive  General Entertainment   
34                                  Roundabout Canada  General Entertainment   
35                               The Reservoir Lounge              Jazz Club   
36                                     D.W. Alexander           Cocktail Bar   
37                                Scotiabank Theatres          Movie Theater   
38                                       The Ballroom          Bowling Alley   
39                             Azure Restaurant & Bar             Restaurant   
40                                               CUBE              Nightclub   
41                               Bier Markt Esplanade     Belgian Restaurant   

                                address  \
0                        225 King St. W   
1                    145 Richmond St. W   
2                          189 Yonge St   
3                        300 King St. W   
4                        301 Front St W   
5                         260 King St W   
6                         91 King St. E   
7                       145 Queen St. W   
8                        350 King St. W   
9                             40 Bay St   
10                     221 King St West   
11                        40 Bay Street   
12  270 Adelaide Street West, Suite 202   
13    

In [53]:
#Understanding types of entertainment venues available in Toronto

Toronto_unique_entertainment= dataframe_filtered_TSX_fun_1['categories'].unique()

### 3.4.5. Creating a Frequency Table for the Data points of interests from the Results of Entertainment spots around the Toronto Stock Exchange 

In [54]:
TSX_entertainment_categories =  dataframe_filtered_TSX_fun_1.groupby('categories')['categories'].size().reset_index(name='counts').sort_values(['counts','categories'],ascending = [False,True])
TSX_entertainment_categories

categories  counts
7   General Entertainment      25
10          Movie Theater       3
0                     Bar       2
1      Basketball Stadium       1
2      Belgian Restaurant       1
3           Bowling Alley       1
4            Cocktail Bar       1
5            Concert Hall       1
6             Event Space       1
8               Jazz Club       1
9     Monument / Landmark       1
11              Nightclub       1
12                   Park       1
13             Restaurant       1
14                Theater       1

# 3.5 Exploring Number of Restaurants around the Financial Districts

### 3.5.1. Exploring Restaurants around the Johannesburg Financial District (Johannesburg Stock Exchange)

In [55]:
#Reducing the radius to 500 meters based on the fact that food spots should not be sparsely located as a result of the smaller footprint required to set them up
query = 'Restaurant'
radius_1 = 500
print(query + ' .... OK!')

Restaurant .... OK!


In [56]:
LIMIT_2 = 500

In [57]:
url_JSE_food = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude, longitude,ACCESS_TOKEN, VERSION, query, radius_1, LIMIT_2)
url_JSE_food

'https://api.foursquare.com/v2/venues/explore?client_id=XP5V0UWY5UNHLA1PLLNKBCGTZCVKXNQIMFLB5BCTUMN3TXMC&client_secret=C0FGSOMMZL20PS3GZJ24N54TVVQAKMG1HKA1GL1YMPLMQPUU&ll=-26.1021936,28.0565306&oauth_token=1UUHMKB3ZBYQ3GVUNHWDRV3VFWZPBOT4YEWSTCO0EMXFYW2N&v=20180604&query=Restaurant&radius=500&limit=500'

### 3.5.2. Formatting Results for the Restaurants around the Johannesburg Financial District (Johannesburg Stock Exchange) API

In [58]:
results_JSE_food = requests.get(url_JSE_food).json()
results_JSE_food

{'meta': {'code': 200, 'requestId': '60dd2ce41c8b5132e9dde8ab'},
 'notifications': [{'type': 'notificationTray', 'item': {'unreadCount': 0}}],
 'response': {'headerLocation': 'City of Johannesburg',
  'headerFullLocation': 'City of Johannesburg',
  'headerLocationGranularity': 'city',
  'query': 'restaurant',
  'totalResults': 34,
  'suggestedBounds': {'ne': {'lat': -26.097693595499994,
    'lng': 28.061532324720723},
   'sw': {'lat': -26.106693604500006, 'lng': 28.051528875279274}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '5b79a0f856837a001cbf334c',
       'name': 'Saint',
       'location': {'lat': -26.10346672778748,
        'lng': 28.059072452559565,
        'labeledLatLngs': [{'label': 'display',
          'lat': -26.10346672778748,
          'lng': 28.059072

In [59]:
'There are {} Restaurants around the Johannesburg Stock Exchange.'.format(len(results_JSE_food['response']['groups'][0]['items']))

'There are 34 Restaurants around the Johannesburg Stock Exchange.'

In [60]:
items_JSE_food = results_JSE_food['response']['groups'][0]['items']
items_JSE_food[0]

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '5b79a0f856837a001cbf334c',
  'name': 'Saint',
  'location': {'lat': -26.10346672778748,
   'lng': 28.059072452559565,
   'labeledLatLngs': [{'label': 'display',
     'lat': -26.10346672778748,
     'lng': 28.059072452559565}],
   'distance': 290,
   'postalCode': '2196',
   'cc': 'ZA',
   'city': 'Sandton',
   'state': 'IGauteng',
   'country': 'iNingizimu Afrika',
   'formattedAddress': ['Sandton', '2196']},
  'categories': [{'id': '4bf58dd8d48988d110941735',
    'name': 'Italian Restaurant',
    'pluralName': 'Italian Restaurants',
    'shortName': 'Italian',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/italian_',
     'suffix': '.png'},
    'primary': True}],
  'photos': {'count': 0, 'groups': []}},
 'referralId': 'e-0-5b79a0f856837a001cbf334c-0'}

In [61]:
dataframe_JSE_food = pd.json_normalize(items_JSE_food) # flatten JSON

### 3.5.3. Displaying Results for the Restaurants around the Johannesburg Financial District (Johannesburg Stock Exchange) API

In [62]:
dataframe_JSE_food.head()

referralId  reasons.count  \
0  e-0-5b79a0f856837a001cbf334c-0              0   
1  e-0-4c2cc3cd2219c928209ba648-1              0   
2  e-0-4be68e7ad4f7c9b63ddf2620-2              0   
3  e-0-4ba8ff50f964a520ea013ae3-3              0   
4  e-0-54967e59498e349305cfdbb1-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id      venue.name  venue.location.lat  \
0  5b79a0f856837a001cbf334c           Saint          -26.103467   
1  4c2cc3cd2219c928209ba648  Rocket Express          -26.103396   
2  4be68e7ad4f7c9b63ddf2620         Pigalle          -26.106063   
3  4ba8ff50f964a520ea013ae3    The Bull Run          -26.103710   
4  54967e59498e349305cfdbb1    Knead Bakery          -26.106008   

   venue.location.lng                      venue.location.labeledLatLngs  \
0           28.059072  [{'label': 'display', 'lat': -26.1034667277874...   
1           28.058821  [{'label': 'display', 'lat': -26.1033958282770...   
2           28.054277  [{'label': 'display', 'lat': -26.1060632898311...   
3           28.058277  [{'label': 'display', 'lat': -26.1037099494713...   
4           28.054516  [{'label': 'display', 'lat': -26.1060075716839...   

   venue.location.distance venue.location.postalCode venue.location.cc  \
0                      290                      2196                ZA   
1                      265                       NaN                ZA   
2                      486                       NaN                ZA   
3                      242                       NaN                ZA   
4                      469                       NaN                ZA   

  venue.location.city venue.location.state venue.location.country  \
0             Sandton             IGauteng      iNingizimu Afrika   
1             Sandton             IGauteng      iNingizimu Afrika   
2             Sandton             IGauteng      iNingizimu Afrika   
3             Sandton             IGauteng      iNingizimu Afrika   
4             Sandton             IGauteng      iNingizimu Afrika   

         venue.location.formattedAddress  \
0                        [Sandton, 2196]   
1                [Maude Street, Sandton]   
2                              [Sandton]   
3  [Maude Street (West Street), Sandton]   
4                [Maude Street, Sandton]   

                                    venue.categories  venue.photos.count  \
0  [{'id': '4bf58dd8d48988d110941735', 'name': 'I...                   0   
1  [{'id': '4bf58dd8d48988d1bd941735', 'name': 'S...                   0   
2  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...                   0   
3  [{'id': '4bf58dd8d48988d1cc941735', 'name': 'S...                   0   
4  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...                   0   

  venue.photos.groups venue.location.address venue.location.crossStreet  
0                  []                    NaN                        NaN  
1                  []           Maude Street                        NaN  
2                  []                    NaN                        NaN  
3                  []           Maude Street                West Street  
4                  []           Maude Street                        NaN

In [63]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe_JSE_food.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered_JSE_food = dataframe_JSE_food.loc[:, filtered_columns]


# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered_JSE_food['venue.categories'] = dataframe_filtered_JSE_food.apply(get_category_type, axis=1)


# clean columns
dataframe_filtered_JSE_food.columns = [col.split('.')[-1] for col in dataframe_filtered_JSE_food.columns]

dataframe_filtered_JSE_food.head(10)

name          categories  \
0                                              Saint  Italian Restaurant   
1                                     Rocket Express         Salad Place   
2                                            Pigalle          Restaurant   
3                                       The Bull Run          Steakhouse   
4                                       Knead Bakery              Bakery   
5                                The Panini Brothers                Café   
6                             Europa Express Nedbank                Café   
7  The Gourmet Garage Sandton burgers, steak, rib...        Burger Joint   
8                                     Fego Caffe RMB                Café   
9                                       Squires Loft          Steakhouse   

         lat        lng                                     labeledLatLngs  \
0 -26.103467  28.059072  [{'label': 'display', 'lat': -26.1034667277874...   
1 -26.103396  28.058821  [{'label': 'display', 'lat': -26.1033958282770...   
2 -26.106063  28.054277  [{'label': 'display', 'lat': -26.1060632898311...   
3 -26.103710  28.058277  [{'label': 'display', 'lat': -26.1037099494713...   
4 -26.106008  28.054516  [{'label': 'display', 'lat': -26.1060075716839...   
5 -26.101723  28.057038  [{'label': 'display', 'lat': -26.1017232847815...   
6 -26.102199  28.058406  [{'label': 'display', 'lat': -26.1021987008985...   
7 -26.102423  28.057541  [{'label': 'display', 'lat': -26.1024230020549...   
8 -26.100613  28.056982  [{'label': 'display', 'lat': -26.100613, 'lng'...   
9 -26.102299  28.056577  [{'label': 'display', 'lat': -26.102299, 'lng'...   

   distance postalCode  cc     city     state            country  \
0       290       2196  ZA  Sandton  IGauteng  iNingizimu Afrika   
1       265        NaN  ZA  Sandton  IGauteng  iNingizimu Afrika   
2       486        NaN  ZA  Sandton  IGauteng  iNingizimu Afrika   
3       242        NaN  ZA  Sandton  IGauteng  iNingizimu Afrika   
4       469        NaN  ZA  Sandton  IGauteng  iNingizimu Afrika   
5        72        NaN  ZA  Sandton  IGauteng  iNingizimu Afrika   
6       187        NaN  ZA  Sandton  IGauteng  iNingizimu Afrika   
7       104       2196  ZA  Sandton  IGauteng  iNingizimu Afrika   
8       181       2146  ZA  Sandton  IGauteng  iNingizimu Afrika   
9        12        NaN  ZA  Sandton  IGauteng  iNingizimu Afrika   

                                    formattedAddress  \
0                                    [Sandton, 2196]   
1                            [Maude Street, Sandton]   
2                                          [Sandton]   
3              [Maude Street (West Street), Sandton]   
4                            [Maude Street, Sandton]   
5  [Shop 6, 24 Cenbral (Gwen Lane and Fredman Dri...   
6                                          [Sandton]   
7  [24 central, Gwen Lane (Fredman Drive and Gwen...   
8  [Shop G01, RMB 2 Merchant Place, 1 Fredman Dri...   
9                              [24 Central, Sandton]   

                                           address  \
0                                              NaN   
1                                     Maude Street   
2                                              NaN   
3                                     Maude Street   
4                                     Maude Street   
5                               Shop 6, 24 Cenbral   
6                                              NaN   
7                            24 central, Gwen Lane   
8  Shop G01, RMB 2 Merchant Place, 1 Fredman Drive   
9                                       24 Central   

                   crossStreet                        id  
0                          NaN  5b79a0f856837a001cbf334c  
1                          NaN  4c2cc3cd2219c928209ba648  
2                          NaN  4be68e7ad4f7c9b63ddf2620  
3                  West Street  4ba8ff50f964a520ea013ae3  
4                          NaN  54967e59498e349305cfdbb1  
5  Gwen Lane and Fredman Drive  51da6

### 3.5.4. Displaying data points of interest from the Data exploration of Restaurants around the Johannesburg Financial District (Johannesburg Stock Exchange) API

In [64]:
dataframe_JSE_food = dataframe_filtered_JSE_food.iloc [:, 0:8]

In [65]:
dataframe_JSE_food

name            categories  \
0                                               Saint    Italian Restaurant   
1                                      Rocket Express           Salad Place   
2                                             Pigalle            Restaurant   
3                                        The Bull Run            Steakhouse   
4                                        Knead Bakery                Bakery   
5                                 The Panini Brothers                  Café   
6                              Europa Express Nedbank                  Café   
7   The Gourmet Garage Sandton burgers, steak, rib...          Burger Joint   
8                                      Fego Caffe RMB                  Café   
9                                        Squires Loft            Steakhouse   
10                                     Gourmet Garage          Burger Joint   
11                                                 HQ            Steakhouse   
12                           4 merchant place canteen            Restaurant   
13                                  Tivoli Restaurant    Italian Restaurant   
14                         Riverside Caffe And Bistro                  Café   
15                                         The Colony        Breakfast Spot   
16                                         Wok N Roll      Asian Restaurant   
17                                 Tashas At The Marc            Restaurant   
18                                    Ribs And Burger          Burger Joint   
19                                     The Blues Room            Restaurant   
20                              The Colony, Balalaika        Breakfast Spot   
21                             Trade Winds Restaurant    African Restaurant   
22                       Lotus Teppanyaki & Sushi Bar            Restaurant   
23                                     Kauai The Marc  Fast Food Restaurant   
24                                   Flavours on west             Cafeteria   
25                          Jewel of india restaurant     Indian Restaurant   
26                              Discovery Snack Shack           Snack Place   
27                                       The Chairman            Restaurant   
28                              Tradewinds Restaurant     Indian Restaurant   
29                Sandton City, Johannesburg, Gauteng                  Café   
30                             Wall Street Restaurant    African Restaurant   
31                                             Subway        Sandwich Place   
32                                           Annica's            Restaurant   
33                        Fego Caffe Discovery Health            Restaurant   

          lat        lng                                     labeledLatLngs  \
0  -26.103467  28.059072  [{'label': 'display', 'lat': -26.1034667277874...   
1  -26.103396  28.058821  [{'label': 'display', 'lat': -26.1033958282770...   
2  -26.106063  28.054277  [{'label': 'display', 'lat': -26.1060632898311...   
3  -26.103710  28.058277  [{'label': 'display', 'lat': -26.1037099494713...   
4  -26.106008  28.054516  [{'label': 'display', 'lat': -26.1060075716839...   
5  -26.101723  28.057038  [{'label': 'display', 'lat': -26.1017232847815...   
6  -26.102199  28.058406  [{'label': 'display', 'lat': -26.1021987008985...   
7  -26.102423  28.057541  [{'label': 'display', 'lat': -26.1024230020549...   
8  -26.100613  28.056982  [{'label': 'display', 'lat': -26.100613, 'lng'...   
9  -26.102299  28.056577  [{'label': 'display', 'lat': -26.102299, 'lng'...   
10 -26.102236  28.056795  [{'label': 'display', 'lat': -26.1022357870133...   
11 -26.102275  28.056616  [{'label': 'display', 'lat': -26.1022745203470...   
12 -26.100307  28.057696  [{'label': 'display', 'lat': -26.1003072759883...   
13 -26.103710  28.058464  [{'label': 'display', 'lat': -26.1037097236843...   
14 -26.103692  28.058530  [{'label': 'display', 'lat': -26.1036924613242...   
15 -26.104065  28.0

### 3.5.5. Creating a Frequency table for the data exploration of Restaurants around the Johannesburg Financial District (Johannesburg Stock Exchange) API

In [66]:
dataframe_JSE_food ['categories'].unique()

array(['Italian Restaurant', 'Salad Place', 'Restaurant', 'Steakhouse',
       'Bakery', 'Café', 'Burger Joint', 'Breakfast Spot',
       'Asian Restaurant', 'African Restaurant', 'Fast Food Restaurant',
       'Cafeteria', 'Indian Restaurant', 'Snack Place', 'Sandwich Place'],
      dtype=object)

In [67]:
dataframe_JSE_food_categories =  dataframe_JSE_food.groupby('categories')['categories'].size().reset_index(name='counts').sort_values(['counts','categories'],ascending = [False,True])
dataframe_JSE_food_categories 

categories  counts
10            Restaurant       8
6                   Café       5
4           Burger Joint       3
14            Steakhouse       3
0     African Restaurant       2
3         Breakfast Spot       2
8      Indian Restaurant       2
9     Italian Restaurant       2
1       Asian Restaurant       1
2                 Bakery       1
5              Cafeteria       1
7   Fast Food Restaurant       1
11           Salad Place       1
12        Sandwich Place       1
13           Snack Place       1

## 3.6 Exploring Restaurants around the Toronto Financial District (Toronto Stock Exchange)

### 3.6.1 Running the API to explore Restaurants around the Toronto Stock Exchange

In [68]:
url_TSX_food = 'https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&ll={},{}&oauth_token={}&v={}&query={}&radius={}&limit={}'.format(CLIENT_ID, CLIENT_SECRET, latitude_2, longitude_2,ACCESS_TOKEN, VERSION, query, radius_1, LIMIT_2)
url_TSX_food

'https://api.foursquare.com/v2/venues/explore?client_id=XP5V0UWY5UNHLA1PLLNKBCGTZCVKXNQIMFLB5BCTUMN3TXMC&client_secret=C0FGSOMMZL20PS3GZJ24N54TVVQAKMG1HKA1GL1YMPLMQPUU&ll=43.64855,-79.383304&oauth_token=1UUHMKB3ZBYQ3GVUNHWDRV3VFWZPBOT4YEWSTCO0EMXFYW2N&v=20180604&query=Restaurant&radius=500&limit=500'

### 3.6.2 Formatting the API results for exploring Restaurants around the Toronto Stock Exchange

In [69]:
results_TSX_food = requests.get(url_TSX_food).json()
results_TSX_food

{'meta': {'code': 200, 'requestId': '60dd2d00a770ad00183f9f50'},
 'notifications': [{'type': 'notificationTray', 'item': {'unreadCount': 0}}],
 'response': {'suggestedFilters': {'header': 'Tap to show:',
   'filters': [{'name': 'Open now', 'key': 'openNow'}]},
  'headerLocation': 'Financial District',
  'headerFullLocation': 'Financial District, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'query': 'restaurant',
  'totalResults': 166,
  'suggestedBounds': {'ne': {'lat': 43.6530500045, 'lng': -79.37709659032932},
   'sw': {'lat': 43.6440499955, 'lng': -79.38951140967067}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4ad69511f964a520e40721e3',
       'name': 'The Keg Steakhouse + Bar - York Street',
       'location': {'address': '165 York St',
        'c

In [70]:
'There are {} Restaurants around the Toronto Stock Exchange.'.format(len(results_TSX_food['response']['groups'][0]['items']))

'There are 100 Restaurants around the Toronto Stock Exchange.'

In [71]:
items_TSX_food = results_TSX_food['response']['groups'][0]['items']
items_TSX_food[0]

{'reasons': {'count': 0,
  'items': [{'summary': 'This spot is popular',
    'type': 'general',
    'reasonName': 'globalInteractionReason'}]},
 'venue': {'id': '4ad69511f964a520e40721e3',
  'name': 'The Keg Steakhouse + Bar - York Street',
  'location': {'address': '165 York St',
   'crossStreet': 'btwn Richmond St. & Adelaide St.',
   'lat': 43.64998659318569,
   'lng': -79.38410336664538,
   'labeledLatLngs': [{'label': 'display',
     'lat': 43.64998659318569,
     'lng': -79.38410336664538}],
   'distance': 172,
   'postalCode': 'M5H 3R8',
   'cc': 'CA',
   'city': 'Toronto',
   'state': 'ON',
   'country': 'Canada',
   'formattedAddress': ['165 York St (btwn Richmond St. & Adelaide St.)',
    'Toronto ON M5H 3R8',
    'Canada']},
  'categories': [{'id': '4bf58dd8d48988d1c4941735',
    'name': 'Restaurant',
    'pluralName': 'Restaurants',
    'shortName': 'Restaurant',
    'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/default_',
     'suffix': '.png'},
    'prim

In [72]:
dataframe_TSX_food = pd.json_normalize(items_TSX_food) # flatten JSON

### 3.6.3 Displaying the API results for exploring Restaurants around the Toronto Stock Exchange

In [73]:
dataframe_TSX_food.head()

referralId  reasons.count  \
0  e-0-4ad69511f964a520e40721e3-0              0   
1  e-0-4dcbf219d22d7ffe9d39197e-1              0   
2  e-0-4ad4c05df964a52059f620e3-2              0   
3  e-0-559a8f5a498e31f945041245-3              0   
4  e-0-5b6c842bc36588002c80a934-4              0   

                                       reasons.items  \
0  [{'summary': 'This spot is popular', 'type': '...   
1  [{'summary': 'This spot is popular', 'type': '...   
2  [{'summary': 'This spot is popular', 'type': '...   
3  [{'summary': 'This spot is popular', 'type': '...   
4  [{'summary': 'This spot is popular', 'type': '...   

                   venue.id                              venue.name  \
0  4ad69511f964a520e40721e3  The Keg Steakhouse + Bar - York Street   
1  4dcbf219d22d7ffe9d39197e                     Brick Street Bakery   
2  4ad4c05df964a52059f620e3                                   Canoe   
3  559a8f5a498e31f945041245                                   Maman   
4  5b6c842bc36588002c80a934                            Cafe Landwer   

  venue.location.address        venue.location.crossStreet  \
0            165 York St  btwn Richmond St. & Adelaide St.   
1          100 King St W                        at Bay St.   
2  66 Wellington St West                     at Bay Street   
3          100 King St W                btwn Bay & York st   
4     165 University Ave                     Adelaide St W   

   venue.location.lat  venue.location.lng  \
0           43.649987          -79.384103   
1           43.648815          -79.380605   
2           43.647452          -79.381320   
3           43.648309          -79.382253   
4           43.648753          -79.385367   

                       venue.location.labeledLatLngs  ...  venue.location.cc  \
0  [{'label': 'display', 'lat': 43.64998659318569...  ...                 CA   
1  [{'label': 'display', 'lat': 43.6488151631615,...  ...                 CA   
2  [{'label': 'display', 'lat': 43.64745206618347...  ...                 CA   
3  [{'label': 'display', 'lat': 43.64830873804684...  ...                 CA   
4  [{'label': 'display', 'lat': 43.64875349932059...  ...                 CA   

  venue.location.city venue.location.state venue.location.country  \
0             Toronto                   ON                 Canada   
1             Toronto                   ON                 Canada   
2             Toronto                   ON                 Canada   
3             Toronto                   ON                 Canada   
4             Toronto                   ON                 Canada   

                     venue.location.formattedAddress  \
0  [165 York St (btwn Richmond St. & Adelaide St....   
1   [100 King St W (at Bay St.), Toronto ON, Canada]   
2  [66 Wellington St West (at Bay Street), Toront...   
3  [100 King St W (btwn Bay & York st), Toronto O...   
4  [165 University Ave (Adelaide St W), Toronto O...   

                                    venue.categories venue.photos.count  \
0  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...                  0   
1  [{'id': '4bf58dd8d48988d16a941735', 'name': 'B...                  0   
2  [{'id': '4bf58dd8d48988d1c4941735', 'name': 'R...                  0   
3  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...                  0   
4  [{'id': '4bf58dd8d48988d16d941735', 'name': 'C...                  0   

  venue.photos.groups  venue.venuePage.id venue.location.neighborhood  
0                  []          1359966175                         NaN  
1                  []                 NaN                         NaN  
2                  []            40552945                         NaN  
3                  []                 NaN                         NaN  
4                  []                 NaN                         NaN  

[5 rows x 22 columns]

In [74]:
# filter columns
filtered_columns = ['venue.name', 'venue.categories'] + [col for col in dataframe_TSX_food.columns if col.startswith('venue.location.')] + ['venue.id']
dataframe_filtered_TSX_food = dataframe_TSX_food.loc[:, filtered_columns]


# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

# filter the category for each row
dataframe_filtered_TSX_food['venue.categories'] = dataframe_filtered_TSX_food.apply(get_category_type, axis=1)


# clean columns
dataframe_filtered_TSX_food.columns = [col.split('.')[-1] for col in dataframe_filtered_TSX_food.columns]

dataframe_filtered_TSX_food.head(10)

name          categories  \
0  The Keg Steakhouse + Bar - York Street          Restaurant   
1                     Brick Street Bakery              Bakery   
2                                   Canoe          Restaurant   
3                                   Maman                Café   
4                            Cafe Landwer                Café   
5             Sam James Coffee Bar (SJCB)                Café   
6                                  Bymark          Restaurant   
7                John & Sons Oyster House  Seafood Restaurant   
8                Astarté Fresh Yogurt Bar                Café   
9                       Pizzeria Libretto         Pizza Place   

                 address                       crossStreet        lat  \
0            165 York St  btwn Richmond St. & Adelaide St.  43.649987   
1          100 King St W                        at Bay St.  43.648815   
2  66 Wellington St West                     at Bay Street  43.647452   
3          100 King St W                btwn Bay & York st  43.648309   
4     165 University Ave                     Adelaide St W  43.648753   
5         150 King St. W                       in the PATH  43.647881   
6    66 Wellington St. W               btwn York & Bay St.  43.647217   
7   56 Temperance Street                               NaN  43.650656   
8   150 King Street West      The PATH @ King & University  43.647596   
9     155 University Ave                       at Pearl St  43.648334   

         lng                                     labeledLatLngs  distance  \
0 -79.384103  [{'label': 'display', 'lat': 43.64998659318569...       172   
1 -79.380605  [{'label': 'display', 'lat': 43.6488151631615,...       219   
2 -79.381320  [{'label': 'display', 'lat': 43.64745206618347...       201   
3 -79.382253  [{'label': 'display', 'lat': 43.64830873804684...        88   
4 -79.385367  [{'label': 'display', 'lat': 43.64875349932059...       167   
5 -79.384332  [{'label': 'display', 'lat': 43.64788137014028...       111   
6 -79.381252  [{'label': 'display', 'lat': 43.64721728401800...       222   
7 -79.381613  [{'label': 'display', 'lat': 43.65065642537144...       271   
8 -79.386419  [{'label': 'display', 'lat': 43.64759591780736...       272   
9 -79.385111  [{'label': 'display', 'lat': 43.64833354686235...       147   

  postalCode  cc     city state country  \
0    M5H 3R8  CA  Toronto    ON  Canada   
1        NaN  CA  Toronto    ON  Canada   
2    M5K 1H6  CA  Toronto    ON  Canada   
3    M5X 1C1  CA  Toronto    ON  Canada   
4    M5H 3B8  CA  Toronto    ON  Canada   
5    M5H 4B6  CA  Toronto    ON  Canada   
6    M5K 1J3  CA  Toronto    ON  Canada   
7    M5H 3V5  CA  Toronto    ON  Canada   
8    M5H 1J9  CA  Toronto    ON  Canada   
9    M5H 3B7  CA  Toronto    ON  Canada   

                                    formattedAddress neighborhood  \
0  [165 York St (btwn Richmond St. & Adelaide St....          NaN   
1   [100 King St W (at Bay St.), Toronto ON, Canada]          NaN   
2  [66 Wellington St West (at Bay Street), Toront...          NaN   
3  [100 King St W (btwn Bay & York st), Toronto O...          NaN   
4  [165 University Ave (Adelaide St W), Toronto O...          NaN   
5  [150 King St. W (in the PATH), Toronto ON M5H ...          NaN   
6  [66 Wellington St. W (btwn York & Bay St.), To...          NaN   
7  [56 Temperance Street, Toronto ON M5H 3V5, Can...          NaN   
8  [150 King Street West (The PATH @ King & Unive...          NaN   
9  [155 University Ave (at Pearl St), Toronto ON ...          NaN   

                         id  
0  4ad69511f964a520e40721e3  
1  4dcbf219d22d7ffe9d39197e  
2  4ad4c05df964a52059f620e3  
3  559a8f5a498e31f945041245  
4  5b6c842bc36588002c80a934  
5  4fccaa8fe4b05a98df3d9417  
6  4ae7199ff964a52068a821e3  
7  50ecb1fae4b0beb13294f0aa  
8  5227fd9d7e48b6c9c707f4fa  
9  5462ac56498e128ccafe8fea

### 3.6.4 Displaying API results Points of Interest from exploring Restaurants around the Toronto Stock Exchange

In [75]:
dataframe_TSX_food = dataframe_filtered_TSX_food.iloc [:, 0:8]

In [76]:
dataframe_TSX_food

name            categories  \
0   The Keg Steakhouse + Bar - York Street            Restaurant   
1                      Brick Street Bakery                Bakery   
2                                    Canoe            Restaurant   
3                                    Maman                  Café   
4                             Cafe Landwer                  Café   
..                                     ...                   ...   
95                                 Wendy’s  Fast Food Restaurant   
96                        Modus Ristorante    Italian Restaurant   
97                                  Subway            Restaurant   
98                            Sunset Grill        Breakfast Spot   
99                   Lee Chen Asian Bistro      Asian Restaurant   

                  address                       crossStreet        lat  \
0             165 York St  btwn Richmond St. & Adelaide St.  43.649987   
1           100 King St W                        at Bay St.  43.648815   
2   66 Wellington St West                     at Bay Street  43.647452   
3           100 King St W                btwn Bay & York st  43.648309   
4      165 University Ave                     Adelaide St W  43.648753   
..                    ...                               ...        ...   
95   55 University Avenue                               NaN  43.646660   
96          145 King St W                              York  43.647621   
97   60 Yonge St, Unit 65                  at Wellington St  43.648300   
98      1 Richmond St. W.                      at Yonge St.  43.651662   
99          214 King St W                         Simcoe St  43.647010   

          lng                                     labeledLatLngs  distance  
0  -79.384103  [{'label': 'display', 'lat': 43.64998659318569...       172  
1  -79.380605  [{'label': 'display', 'lat': 43.6488151631615,...       219  
2  -79.381320  [{'label': 'display', 'lat': 43.64745206618347...       201  
3  -79.382253  [{'label': 'display', 'lat': 43.64830873804684...        88  
4  -79.385367  [{'label': 'display', 'lat': 43.64875349932059...       167  
..        ...                                                ...       ...  
95 -79.383878  [{'label': 'display', 'lat': 43.6466603, 'lng'...       215  
96 -79.383350  [{'label': 'display', 'lat': 43.64762079256355...       103  
97 -79.377846  [{'label': 'display', 'lat': 43.64830041435992...       440  
98 -79.379075  [{'label': 'display', 'lat': 43.65166162378037...       485  
99 -79.386536  [{'label': 'display', 'lat': 43.64700977986747...       311  

[100 rows x 8 columns]

### 3.6.5 Displaying Frequency results from the API results for exploring Restaurants around the Toronto Stock Exchange

In [77]:
dataframe_TSX_food_categories =  dataframe_TSX_food.groupby('categories')['categories'].size().reset_index(name='counts').sort_values(['counts','categories'],ascending = [False,True])
dataframe_TSX_food_categories 

categories  counts
26                     Restaurant      14
8                            Café       9
1                Asian Restaurant       5
11                  Deli / Bodega       5
12           Fast Food Restaurant       5
28                 Sandwich Place       5
3                          Bakery       4
20            Japanese Restaurant       4
32               Sushi Restaurant       4
6                    Burger Joint       3
7                   Burrito Place       3
19             Italian Restaurant       3
27                    Salad Place       3
31                     Steakhouse       3
34  Vegetarian / Vegan Restaurant       3
0             American Restaurant       2
5                  Breakfast Spot       2
9              Chinese Restaurant       2
13                     Food Court       2
15                      Gastropub       2
24                    Pizza Place       2
33                Thai Restaurant       2
2                      Bagel Shop       1
4            Brazilian Restaurant       1
10           Colombian Restaurant       1
14            Fried Chicken Joint       1
16         Gluten-free Restaurant       1
17               Greek Restaurant       1
18              Indian Restaurant       1
21       Mediterranean Restaurant       1
22             Mexican Restaurant       1
23        New American Restaurant       1
25               Ramen Restaurant       1
29             Seafood Restaurant       1
30                     Soup Place       1

# 4. Discussion


Buildings around Financial Centres
Explorations of the buildings around the Johannesburg Stock Exchange revealed 6 buildings of interest with 66% of them being Hotels and the remaining buildings being a shopping Mall and Wall Street restaurant. Exploration the buildings around the Toronto Stock exchange revealed 65 buildings with diverse categories of which only 12% of them was hotels. Though in terms of percentage Toronto’s number is low in terms of quantity it is still more than Johannesburg’s. Toronto also offer 41 more categories that I was not able to find in the Johannesburg data. 

Exploration of entertainment areas around Johannesburg only reveal 4 spots with only two categories being a nightclub and general entertainment. For the same category Toronto offers 42 areas of entertainments, offering visitors variety of things to do and explore around the Financial District. In terms of the categories of entertainment Toronto has 25 (60%) venues categorised under general entertainment. 40% of offer alternative categories which are not found around the Johannesburg Financial district area. 

Exploration to the food market in both cities, indicate that there are 34 restaurant around the Johannesburg stock exchange  compared to more than 100 around the Toronto Stock exchange. There is 32 categories of restaurants in the Toronto Stock exchange area compared to 15 around the Johannesburg stock exchange. Though Johannesburg offers fewer restaurants there is more variety in this category for visitors to the area than in any other category explored in this study. 


# 5. Recommendation


The data does indicate low levels of maturity especially for the Johannesburg data where there were no Office building categories which is a little hard to fathom considering that Johannesburg is the Financial hub of the country of South Africa. As such, I do recommend adding that the provider finds a way to get more data points. 

The City of Toronto has more categories of buildings and entertainment areas indicating that there is more points of interest for both visitors and investors. This number of interests do indicate that people are likely to spend in the neighbourhood. Recommendation would be for Johannesburg to invest in more diverse buildings so as to create more interest in the city for people to spend more on the local economy. 

# 6. Conclusion


Toronto does offer variety of areas of interests, entertainment, and restaurants compared to Johannesburg. People visiting the Toronto exchange neighbourhood are likely to spend more time in the area as well as their money compared to Johannesburg. Opportunities of enriched experience are therefore more in the Toronto area compared to Johannesburg. 
